### PR TITLE
Rewrite Parser Using Alex and Happy

### PR DESCRIPTION
**Pull request by solomon-b**
Resolves issues #8 #12 #25.

By rewriting the parser with Happy and Alex we now have accurate span generation, non-hacky string templates, a formalized grammar, and an overall more extensible parser.

I had to remove the QuickCheck Roundtrip tests. They require a span aware pretty printer which we should work on next.

The Golden Tests are a bit of a mess. It is hard to tell what test covers what language feature. I think we should rewrite them in a more organized way. Should that be included in this PR or a followup?

### CODE REVIEW

**Comment by jkachmar** [comment]
@solomon-b Presumably the majority of this PR is generated code; can you do a first-pass and just mark off some areas of relevant files that reviewers should take a look at to help us navigating this a bit?

**Comment by solomon-b** [comment]
> @solomon-b Presumably the majority of this PR is generated code; can you do a first-pass and just mark off some areas of relevant files that reviewers should take a look at to help us navigate this a bit?

Yes I can do that now

**Comment by solomon-b** [comment]
@jkachmar I removed the generated Lexer and Parser modules. Cabal can generate them automatically when building the probject and running the repl. I'm assuming this will work when we import the project as a library as well but I'll need to test that before we merge.

This gets rid of all generated code other then golden test results. I can add comments to key areas of change, but this PR represents a pretty significant rewrite of most of Kriti.

File: src/Kriti/Parser/Lexer.x
Comment (solomon-b) [inline src/Kriti/Parser/Lexer.x:29]: Template lexing is done using [Start Codes](https://www.haskell.org/alex/doc/html/alex-files.html#startcodesl). I tried to document this part of the lexer pretty heavily.

File: src/Kriti/Parser/Monad.hs
Comment (solomon-b) [inline src/Kriti/Parser/Monad.hs:30]: Parsing and Lexing both use the `Parser` Monad.

File: src/Kriti/Parser/Lexer.x
Comment (solomon-b) [inline src/Kriti/Parser/Lexer.x:110]: This is our `Parser Token` monadic wrapper for the [alexScan](https://github.com/simonmar/alex/blob/ab87af1803a5e2f2c09b09eb024dc6ec9f44b0e3/src/Scan.hs#L297) function.

File: src/Kriti/Parser/Grammar.y
Comment (solomon-b) [inline src/Kriti/Parser/Grammar.y:20]: Here is our parser specification. It should be fairly easy to grok, but let me know if you have any specific questions.

File: src/Kriti/Parser/Token.hs
Comment (solomon-b) [inline src/Kriti/Parser/Token.hs:57]: This function is used strictly for rending parse errors and should be replaced by a span aware pretty printer in subsequent PR.

File: src/Kriti/Parser/Lexer.x
Comment (solomon-b) [inline src/Kriti/Parser/Lexer.x:75]: we also use Status Codes for single quoted literals. These are used strictly for `['foo bar']` object lookup syntax.

File: src/Kriti/Parser/Lexer.x
Comment (solomon-b) [inline src/Kriti/Parser/Lexer.x:56]: I get really confused when it comes to handling escape characters in strings and am not sure if this regex is totally correct.

File: src/Kriti.hs
Comment (solomon-b) [inline src/Kriti.hs:11]: Kriti now assumes `ByteStrings` rather then `Text`. Alex only really works with ByteStrings or Strings.

File: test/data/parser/failure/actual/missingClosingBracket.txt
Comment (sordina) [inline test/data/parser/failure/actual/missingClosingBracket.txt:1]: It might be cool in future if we can find a way to render the errors without reference to the implementation - I.e. describe the error and the position, but not reference Alex, etc.

Comment (sordina) [reply test/data/parser/failure/actual/missingClosingBracket.txt:1]: Not a blocker.

**Review by sordina — APPROVED** [review]
Looks fantastic! So much cleaner this way. The only thing I can think of that might be a good addition is some property tests, re. round-trips and evaluation, but that's not a blocker.

File: test/data/parser/failure/actual/missingClosingBracket.txt
Comment (solomon-b) [reply test/data/parser/failure/actual/missingClosingBracket.txt:1]: I agree. I think that would fall under issue #7

**Comment by solomon-b** [comment]
> Looks fantastic! So much cleaner this way. The only thing I can think of that might be a good addition is some property tests, re. round-trips and evaluation, but that's not a blocker.

Thanks! We used to have round trip tests but I had to remove them because of the introduction of Spans into the lexer. I can bring them back once we have a pretty printer (#7).

File: src/Kriti/Parser/Lexer.x
Comment (jkachmar) [reply src/Kriti/Parser/Lexer.x:56]: If you're not sure then it might be worth adding a small `-- NOTE:` comment to that effect, but that's up to you.

File: src/Kriti/Parser/Lexer.x
Comment (jkachmar) [reply src/Kriti/Parser/Lexer.x:110]: This comment looks incomplete; it ends with `We hold the latest`, but it seems like there should be more commentary here.

File: src/Kriti/Parser/Monad.hs
Comment (jkachmar) [inline src/Kriti/Parser/Monad.hs:160]: What's the guidance behind these `INLINE` annotations?

File: src/Kriti/Parser/Monad.hs
Comment (jkachmar) [inline src/Kriti/Parser/Monad.hs:209]: It looks like this is an `encoding` failure, not a `decoding` failure; additionally, the error message should include the aberrant character.

For example:
`error $ "The following character produced an empty UTF8-encoded bytestring, which should be impossible: " ++ [ c ]`

File: src/Kriti/Parser/Token.hs
Comment (jkachmar) [reply src/Kriti/Parser/Token.hs:58]: This is probably worth a `-- NOTE:` comment stating as much, or something.

File: src/Kriti/Parser/Spans.hs
Comment (jkachmar) [inline src/Kriti/Parser/Spans.hs:49]: This could use some documentation.

File: src/Kriti/Parser/Spans.hs
Comment (jkachmar) [inline src/Kriti/Parser/Spans.hs:61]: As a rule, in general, typeclasses should be documented. In this particular case, I'm not 100% sure what `Located` is meant to do, but I can try to intuit that it's meant to recover a `Span` from some arbitrary type.

File: src/Kriti/Parser/Token.hs
Comment (jkachmar) [inline src/Kriti/Parser/Token.hs:4]: The ADTs in this module should have some amount of documentation (`Symbol`, `Token`, `Accessor`, `ValueExt`).

File: test/Spec.hs
Comment (jkachmar) [inline test/Spec.hs:54]:
```suggestion
    it "lexes serialized tokens and yields those tokens modulo spans" $
```
`hspec` tests should read almost like sentences (e.g. "it lexes serialized tokens and yields those tokens modulo spans").

File: src/Kriti/Parser/Monad.hs
Comment (solomon-b) [reply src/Kriti/Parser/Monad.hs:160]: I didn't look at the core or anything but its a good idea to make sure all of this code gets inlined.

File: src/Kriti/Parser/Monad.hs
Comment (solomon-b) [reply src/Kriti/Parser/Monad.hs:209]: good catch!

File: src/Kriti/Parser/Lexer.x
Comment (solomon-b) [reply src/Kriti/Parser/Lexer.x:59]: good call.

**Review by jkachmar — APPROVED** [review]
LGTM :tada:

**Pull request merged**
